// ola-maps/dist/places.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Places = void 0;
var const_1 = require("./const");
var http_client_1 = require("./utilities/http-client");
/**
 * Represents a client for interacting with a Places API.
 * Provides methods for autocomplete and geocoding functionality.
 */
var Places = /** @class */ (function () {
    /**
     * Creates a new Places client.
     * @param apiKey - Your API key for the Places service.
     * @param version - (Optional) The API version to use. Defaults to 'v1'.
     * @throws {Error} If the API key is missing or invalid.
     */
    function Places(apiKey, version) {
        if (version === void 0) { version = "v1"; }
        /**
         * API key for authentication with the Places service.
         * @private
         */
        this.apiKey = "";
        /**
         * The API version to use. Defaults to 'v1'.
         * @private
         */
        this.version = "v1";
        /**
         * The Axios instance used for making HTTP requests.
         * @private
         */
        this.httpClient = {};
        if (!apiKey || apiKey.trim() === "") {
            throw new Error("API Key is required. Check documentation - https://maps.olakrutrim.com/docs/auth");
        }
        this.apiKey = apiKey;
        this.version = version;
        this.init();
    }
    /**
     * Initializes the Places client by setting up the Axios instance and environment variables.
     * @private
     */
    Places.prototype.init = function () {
        var env = const_1.ENV[this.version];
        process.env.API_VERSION = this.version;
        for (var variable in env) {
            if (env[variable]) {
                process.env[variable] = env[variable];
            }
        }
        this.httpClient = (0, http_client_1.httpClientInit)(this.apiKey);
    };
    /**
     * Performs autocomplete based on the provided input and options.
     * @param input - The user's search input.
     * @param options - (Optional) Additional options to customize the request.
     * @returns A Promise that resolves to the autocomplete results.
     */
    Places.prototype.autocomplete = function (input, options) {
        return __awaiter(this, void 0, void 0, function () {
            var path, data, err_1;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 2, , 3]);
                        path = "".concat(process.env.autoCompletePath, "?input=").concat(encodeURI(input));
                        if (options) {
                            if (options["X-Correlation-Id"]) {
                                path += "X-Correlation-Id=".concat(options["X-Correlation-Id"], "&");
                            }
                            if (options["X-Request-Id"]) {
                                path += "X-Request-Id=".concat(options["X-Request-Id"], "&");
                            }
                            if (options.language) {
                                path += "language=".concat(options.language, "&");
                            }
                            if (typeof options.radius === "number" && options.radius >= 0) {
                                path += "radius=".concat(options.radius, "&");
                            }
                            if (typeof options.strictbounds === "boolean") {
                                path += "strictbounds=".concat(options.strictbounds, "&");
                            }
                            if (options.location) {
                                path += "location=".concat(options.location.latitude, ",").concat(options.location.longitude, "&");
                            }
                        }
                        if (path.endsWith("&") || path.endsWith("?")) {
                            path = path.slice(0, -1);
                        }
                        return [4 /*yield*/, this.httpClient.get(path)];
                    case 1:
                        data = (_a.sent()).data;
                        return [2 /*return*/, Promise.resolve(data)];
                    case 2:
                        err_1 = _a.sent();
                        return [2 /*return*/, Promise.reject(err_1)];
                    case 3: return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Performs geocode based on the provided address and options.
     * @param address - The address to geocode.
     * @param options - (Optional) Additional options to customize the request.
     * @returns A Promise that resolves to the geocoding results.
     */
    Places.prototype.geocode = function (address, options) {
        return __awaiter(this, void 0, void 0, function () {
            var path, x, y, data, err_2;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 2, , 3]);
                        path = "".concat(process.env.geoCodingPath, "?address=").concat(encodeURI(address));
                        if (options) {
                            if (options["X-Correlation-Id"]) {
                                path += "X-Correlation-Id=".concat(options["X-Correlation-Id"], "&");
                            }
                            if (options["X-Request-Id"]) {
                                path += "X-Request-Id=".concat(options["X-Request-Id"], "&");
                            }
                            if (options.language) {
                                path += "language=".concat(options.language, "&");
                            }
                            if (options.bounding) {
                                x = [options.bounding.x.latitude, options.bounding.x.longitude].join(",");
                                y = [options.bounding.y.latitude, options.bounding.y.longitude].join(",");
                                path += "bounding=".concat([x, y].join("|"), "&");
                            }
                        }
                        if (path.endsWith("&") || path.endsWith("?")) {
                            path = path.slice(0, -1);
                        }
                        return [4 /*yield*/, this.httpClient.get(path)];
                    case 1:
                        data = (_a.sent()).data;
                        return [2 /*return*/, Promise.resolve(data)];
                    case 2:
                        err_2 = _a.sent();
                        return [2 /*return*/, Promise.reject(err_2)];
                    case 3: return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Performs reverse geocode to get address details from latitude and longitude coordinates.
     *
     * @param lat - The latitude coordinate.
     * @param lng - The longitude coordinate.
     * @param options - (Optional) Additional options to customize the request:
     *   - X-Correlation-Id: (Optional) A correlation ID for tracking the request.
     *   - X-Request-Id: (Optional) A unique request ID.
     * @returns A Promise that resolves to the reverse geocoding results.
     * @throws {Error} If an error occurs during the request.
     */
    Places.prototype.reverse_geocode = function (lat, lng, options) {
        return __awaiter(this, void 0, void 0, function () {
            var path, data, err_3;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 2, , 3]);
                        path = "".concat(process.env.reverseGeoCodingPath, "?latlng=").concat(lat, ",").concat(lng, "?");
                        if (options) {
                            if (options["X-Correlation-Id"]) {
                                path += "X-Correlation-Id=".concat(options["X-Correlation-Id"], "&");
                            }
                            if (options["X-Request-Id"]) {
                                path += "X-Request-Id=".concat(options["X-Request-Id"], "&");
                            }
                        }
                        if (path.endsWith("&") || path.endsWith("?")) {
                            path = path.slice(0, -1);
                        }
                        return [4 /*yield*/, this.httpClient.get(path)];
                    case 1:
                        data = (_a.sent()).data;
                        return [2 /*return*/, Promise.resolve(data)];
                    case 2:
                        err_3 = _a.sent();
                        return [2 /*return*/, Promise.reject(err_3)];
                    case 3: return [2 /*return*/];
                }
            });
        });
    };
    return Places;
}());
exports.Places = Places;
